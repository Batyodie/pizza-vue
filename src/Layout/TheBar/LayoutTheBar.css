@import "../../style/Base/Extend/BaseExtend.css";

.SortingBar {
  @extend .ContentWrapper;
  justify-content: space-between;
}
.TagGroup {
  @extend .ContentWrapper;
  justify-content: flex-start;
  width: initial;
  padding: 0;
  margin: 0;
}
.TagMargin {
  margin: 10px;
}

.IsSelected {
  --bg-tag: black;
  background-color: var(--bg-tag);
  border: var(--bg-tag);

  & span {
    --color-text: white;
    color: var(--color-text);
  }
}
