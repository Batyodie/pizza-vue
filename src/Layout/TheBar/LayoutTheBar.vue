<script>
import TagGroupMobile from "./TagGroupMobile/TagGroupMobile.vue";
import { Button, DropDown } from "@/components";
import { BtnTag, DropDownStyle } from "@/components/style";

import { mapGetters, mapActions } from "vuex";

import { TheBarStyle } from "@/Layout/style";
export default {
  name: "SortingSideBar",
  components: { DropDown, Button, TagGroupMobile },

  data() {
    return {
      DropDownTagsIsOpen: false
    };
  },

  computed: {
    TheBarStyle() {
      return TheBarStyle;
    },

    BtnTag() {
      return BtnTag;
    },

    DropDownStyle() {
      return DropDownStyle;
    },

    ...mapGetters({
      getTags: "getTags",
      activeTag: "getIsActiveTag",
      getState: "getStateDropdown"
    }),

    getSelected() {
      return tagID => {
        return this.activeTag === tagID ? [TheBarStyle.IsSelected] : "";
      };
    }
  },
  methods: {
    ...mapActions({
      selectedTag: "TheBarSelectedTag",
      filterPizzas: "fetchProducts",
      closeGlobal: "DropDownGlobalClosed"
    }),

    getSelectTag(TagIndex) {
      this.selectedTag(TagIndex);
      this.filterPizzas();
    },

    DropDownTagsOpen() {
      this.DropDownTagsIsOpen = !this.DropDownTagsIsOpen;
    }
  }
};
</script>

<template>
  <div :class="TheBarStyle.SortingBar">
    <div v-if="$mq.resize && $mq.above('1300px')" :class="TheBarStyle.TagGroup">
      <Button
        v-for="tag in getTags"
        :key="tag.id"
        :class="[getSelected(tag.id), TheBarStyle.TagMargin, BtnTag.Tag]"
        :tagIndex="tag.id"
        @selectTag="getSelectTag"
      >
        <template slot="ButtonText">
          {{ tag.text }}
        </template>
      </Button>
    </div>

    <div v-else :class="TheBarStyle.TagGroup">
      <TagGroupMobile :getTags="getTags" :activeTag="activeTag" />
    </div>
    <div>
      <DropDown />
    </div>
  </div>
</template>
