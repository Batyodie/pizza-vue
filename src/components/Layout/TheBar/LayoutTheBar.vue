<template>
  <div :class="TheBarStyle.SortingBar">
    <div v-if="$mq.resize && $mq.above('1300px')" :class="TheBarStyle.TagGroup">
      <Button
        v-for="tag in getTags"
        :key="tag.id"
        :class="[
          getIsActiveTag === tag.id ? [TheBarStyle.IsSelected] : '',
          TheBarStyle.TagMargin,
          TagStyle.Tag,
        ]"
        :index="tag.id"
        :TagStyleBody="TagStyle.TagBody"
        @TheBarTagindex="getSelectTagIndex"
        @select="getSelectTag"
      >
        <template slot="ButtonText">
          {{ tag.text }}
        </template>
      </Button>
    </div>

    <div v-else :class="TheBarStyle.TagGroup">
      <div :class="DropDownStyle.DropDown">
        <div style="display: flex; align-items: center;">
          <!-- <img width="10" height="6" :src="icon" alt="Arrow top" /> -->

          <span :class="DropDownStyle.DropDownLabel">
            Сортировка по:
          </span>
          <button @click="DropDownTagsOpen" :class="DropDownStyle.DropDownBtn">
            Категориям
          </button>
        </div>
        <div
          v-show="DropDownTagsIsOpen"
          :class="DropDownStyle.DropDownPopUp"
          style="max-width: 160px"
        >
          <div
            :class="DropDownStyle.DropDownList"
            style="padding-left: 10px; padding-right: 10px;"
          >
            <Button
              v-for="tag in getTags"
              :key="tag.id"
              :class="[
                getIsActiveTag === tag.text ? [TheBarStyle.IsSelected] : '',
                TheBarStyle.TagMargin,
                TagStyle.Tag,
              ]"
              :TagStyleBody="TagStyle.TagBody"
              @select="getSelectTag"
            >
              <template slot="ButtonText">
                {{ tag.text }}
              </template>
            </Button>
          </div>
        </div>
      </div>
    </div>
    <div>
      <DropDown />
    </div>
  </div>
</template>

<script>
import Button from "../../Module/Button/Button.vue";
import DropDown from "../../Module/DropDown/DropDown.vue";
import DropDownStyle from "../../Module/DropDown/DropDown.css";
import { mapGetters, mapActions } from "vuex";
import TagStyle from "../../Module/Tag/Tag.css";
import TheBarStyle from "./LayoutTheBar.css";
export default {
  components: { DropDown, Button },
  name: "SortingSideBar",
  data() {
    return {
      DropDownTagsIsOpen: false,
      TheBarTagIndex: null,
    };
  },
  computed: {
    TheBarStyle() {
      return TheBarStyle;
    },
    TagStyle() {
      return TagStyle;
    },
    DropDownStyle() {
      return DropDownStyle;
    },
    ...mapGetters({
      getTags: "getTags",
      getIsActiveTag: "getIsActiveTag",
      getState: "getStateDropdown",
    }),
  },
  methods: {
    ...mapActions({
      selectedTag: "TheBarSelectedTag",
      sortPizzas: "GetThePizzas",
    }),
    getSelectTag(TagIndex) {
      this.selectedTag(TagIndex);
      this.sortPizzas();
    },
    DropDownTagsOpen() {
      this.DropDownTagsIsOpen = !this.DropDownTagsIsOpen;
    },
    getSelectTagIndex(index) {
      this.TheBarTagIndex = index;
    },
  },
};
</script>
