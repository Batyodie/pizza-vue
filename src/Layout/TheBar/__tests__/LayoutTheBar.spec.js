import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";
import * as vMediaQuery from "v-media-query";
import LayoutTheBar from "@/Layout/TheBar/LayoutTheBar";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(vMediaQuery.default);

describe("Unit test from LayoutTheBar components", () => {
  window.matchMedia =
    window.matchMedia ||
    function() {
      return {
        matches: true,
        addListener: function() {},
        removeListener: function() {}
      };
    };
  // store data
  const tags = [
    { id: null, text: "Все", type: "all" },
    { id: 0, text: "Мясные", type: "meet" },
    { id: 1, text: "Вегетарианская", type: "vengry" },
    { id: 2, text: "Гриль", type: "grill" },
    { id: 3, text: "Острые", type: "sharp" },
    { id: 4, text: "Закрытые", type: "closed" }
  ];
  const activeTag = null;

  const DropDownIsOpen = false;
  const DropDownItems = [
    { id: 5, text: "популярности", type: "rating", order: "desc" },
    { id: 6, text: "по цене", type: "price", order: "desc" },
    { id: 7, text: "по алфавиту", type: "name", order: "asc" }
  ];
  const DropDownItemIsActive = {
    text: "популярности",
    type: "rating",
    order: "desc"
  };
  let store;
  let actions;
  let getters;
  beforeEach(() => {
    actions = {
      TheBarSelectedTag: jest.fn(),
      fetchProducts: jest.fn(),
      DropDownGlobalClosed: jest.fn(),
      DropDownOpened: jest.fn(),
      DropDownSelected: jest.fn()
    };
    getters = {
      getTags: () => tags,
      getIsActiveTag: () => activeTag,
      getStateDropdown: () => {
        return {
          DropDownIsOpen,
          DropDownItems,
          DropDownItemIsActive
        };
      }
    };

    store = new Vuex.Store({
      actions,
      getters
    });
  });
  // create wrapper factory
  const wrapperFactory = (mouning = shallowMount, { propsData } = {}) => {
    return mouning(LayoutTheBar, {
      localVue,
      store,
      propsData: {
        ...propsData
      }
    });
  };
  it("should initialize correctly layout the bar component", () => {
    const wrapper = wrapperFactory(mount);

    // correctly Vue instance
    expect(wrapper.exists()).toBeTruthy();
    // We expect the wrapper to be layout the bar component
    expect(wrapper.is(LayoutTheBar)).toBeTruthy();
    expect(wrapper.vm.DropDownTagsIsOpen).toBe(false);
  });

  it("should correctly render tags in layout the bar component", () => {
    const wrapper = wrapperFactory(mount);
    let index = 0;
    const listTags = wrapper.findAll(".Tag");

    listTags.wrappers.forEach((e, i) => {
      const renderTag = `${tags[index].text}`;
      if (index === i) {
        expect(listTags.at(index).text()).toEqual(renderTag);
        index++;
      }
    });
    expect(listTags.at(0).classes("IsSelected")).toBe(true);
  });

  it("should correctly render dropdown  in layout the bar", () => {
    const wrapper = wrapperFactory(mount);
    const DropDown = wrapper.find(".DropDown");
    let index = 0;
    const renderLabel = `Сортировка по: 
        ${DropDownItems[index].text}`;

    expect(DropDown.text()).toEqual(renderLabel);
  });

  it("should match the snapshot layout the bar component", () => {
    const wrapper = wrapperFactory(mount);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should correctly selection tag in layout the bar component", async () => {
    const wrapper = wrapperFactory(mount);
    let index = 1;
    const listTags = wrapper.findAll(".Tag").at(index);

    await listTags.trigger("click");
    expect(actions.TheBarSelectedTag).toHaveBeenCalled();
    expect(actions.fetchProducts).toHaveBeenCalled();
  });

  it("should correctly open dropdown in layout the bar", async () => {
    const wrapper = wrapperFactory(mount);

    const DropDown = wrapper.find(".DropDown");
    const DropDownBtn = DropDown.find(".Btn");

    await DropDownBtn.trigger("click");

    expect(actions.DropDownOpened).toHaveBeenCalled();
  });
});
