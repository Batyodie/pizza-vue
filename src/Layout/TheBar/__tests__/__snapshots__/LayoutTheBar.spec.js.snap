// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test from LayoutTheBar components should match the snapshot layout the bar component 1`] = `
<div class="SortingBar">
  <div class="TagGroup"><button type="button" class="Btn IsSelected TagMargin Tag"><span class="Body">
        Все
      </span> </button><button type="button" class="Btn TagMargin Tag"><span class="Body">
        Мясные
      </span> </button><button type="button" class="Btn TagMargin Tag"><span class="Body">
        Вегетарианская
      </span> </button><button type="button" class="Btn TagMargin Tag"><span class="Body">
        Гриль
      </span> </button><button type="button" class="Btn TagMargin Tag"><span class="Body">
        Острые
      </span> </button><button type="button" class="Btn TagMargin Tag"><span class="Body">
        Закрытые
      </span> </button></div>
  <div>
    <div class="DropDown">
      <div style="display: flex; align-items: center;"><img width="10" height="6" src="" alt="Arrow top" class="IconImage"> <span class="Label">Сортировка по:</span> <button class="Btn">
          популярности
        </button></div>
      <!---->
    </div>
  </div>
</div>
`;
